### PR TITLE
#159987722 Add Search Functionality and Custom filtering based on parameters

### PR DESCRIPTION
**Pull request by gonuoha**
### What does this PR do?
add search and filter functionality.

### Description of Task to be completed?
add unit tests for search feature
add search in articles title and description by keywords
add search articles by authors
add search articles by tags

### How should this be manually tested?
Run 'npm install'
create a .env using the .env.sample template
Run 'npm run migrate'
Run 'npm run start:dev'
use postman to test:
get: /api/articles/search?author=someAuthor
get: /api/articles/search?tag=tag1+tag2
get: /api/articles/search?keywords=somethin+in+titlleOrDescription

### What are the relevant pivotal tracker stories?
[#159987722](https://www.pivotaltracker.com/story/show/159987711)

### CODE REVIEW

**Review by emekafredy — COMMENTED** [review]
LGTM. Although you should update the swagger docs with your endpoints

File: server/controllers/article.js [hunk 1]
```diff
@@ -5,6 +5,155 @@ import paginateArticle from '../utils/articlesPaginate';
...
+  }
+
+  /**
+   * @param {list} tag
+   * @param {object} res
+   * @param {object} next
+   * @return {object} jsonResponse
+   */
+  static byTags(tag, res, next) {
+    // trats each word as separate tag
```
Comment (lauragift21) [inline server/controllers/article.js]: Minor typo on your comments.

File: server/controllers/article.js [hunk 2]
```diff
@@ -252,6 +401,33 @@ class ArticleController {
...
+
+  /**
+   * @static
+   * @param {object} req
+   * @param {object} res
+   * @param {object} next
+   * @return {object} jsonResponse
+  */
+  static search(req, res, next) {
+    // expectes author to be author username
```
Comment (lauragift21) [inline server/controllers/article.js]: Same as line 29

**Review by lauragift21 — APPROVED** [review]
LGTM asides the typo I spotted.

File: server/controllers/article.js [hunk 3]
```diff
@@ -5,6 +5,155 @@ import paginateArticle from '../utils/articlesPaginate';
...
+            ]
+          }],
+      attributes: ['id', 'username']
+    }).then(articles => res.status(200).json({
+      status: 'success',
+      articles
+    })).catch(next);
+  }
+}
+
```
Comment (CEOehis) [inline server/controllers/article.js]: Nice work on this implementation of search functionality. However, in lines `11-157`, I think you can make the controllers leaner if you extract the Search class into its' own javascript file and then import its' static methods for use in the `articleController`.

I think, your line of thought was geared toward something like private methods, but I think implementing those methods from the search class as helpers/middleware would provide better separation of concerns.

**Review by CEOehis — CHANGES_REQUESTED** [review]

**Review by CEOehis — APPROVED** [review]

**Review by emekafredy — APPROVED** [review]
LGTM

File: server/controllers/search.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
+import db from '../models';
```
Comment (aknwosu) [inline server/controllers/search.js]: import { Article, User, Tag } from '../models';

File: server/tests/controllers/articles.js [hunk 5]
```diff
@@ -318,5 +318,63 @@ describe('Articles controller', () => {
           done();
         });
     });
+
+    it('should return error if article does not exist', (done) => {
+      chai.request(server)
+        .delete(`/api/articles/${testSlug}e`)
+        .set('Authorization', `Bearer ${token1}`)
+        .end((err, res) => {
+          res.should.have.status(404);
```
Comment (aknwosu) [inline server/tests/controllers/articles.js:328]: Whats the content of the result

Comment (AdmiralOnu) [reply server/tests/controllers/articles.js:328]: this is fixed

File: server/controllers/search.js (hunk 4 above)
Comment (AdmiralOnu) [reply server/controllers/search.js]: this is now fixed, thank you.

**Pull request merged**